### PR TITLE
Added recipe selection UI for Transfer Assembler

### PR DESCRIPTION
**Pull request by Ved-s**
изображение

### CODE REVIEW

File: MechTransfer.cs [hunk 1]
```diff
@@ -365,6 +371,8 @@ public override void AddRecipes()
         public override void PostAddRecipes()
         {
             NetRouter.Init(0);
+
+
```
Comment (DRKV333) [inline MechTransfer.cs]: Remove this empty space.

File: UI/RecipeUI.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        public bool highlight = false;
+
+        protected override void DrawSelf(SpriteBatch spriteBatch)
+        {
+            if (recipe is null) return;
+
+            int ing;
+            for (ing = 0; ing < recipe.requiredItem.Length; ing++)
+                if (recipe.requiredItem[ing].IsAir)
+                    break;
```
Comment (DRKV333) [inline UI/RecipeUI.cs]: For loops should always have braces.

File: Tiles/TransferAssemblerTile.cs [hunk 3]
```diff
@@ -175,6 +178,63 @@ private bool SearchStation(Recipe recipe, int x, int y, out bool alchemy)
...
+                        entity.selectedRecipeHash = 0;
+                        goto StartScrollHandling;
+                    }
+
+                    if (PlayerInput.ScrollWheelDeltaForUI < 0)
+                        index++;
+                    else
+                    {
+                        index--;
+                    }
```
Comment (DRKV333) [inline Tiles/TransferAssemblerTile.cs]: Ifs should have braces on all branches, or none of them.

File: Tiles/TransferAssemblerTile.cs [hunk 4]
```diff
@@ -175,6 +178,63 @@ private bool SearchStation(Recipe recipe, int x, int y, out bool alchemy)
             return true;
         }
 
+        public override void DisplayTooltip(int i, int j)
+        {
+            if (!TryGetEntity(i, j, out TransferAssemblerTileEntity entity)) return;
```
Comment (DRKV333) [inline Tiles/TransferAssemblerTile.cs]: Statements after ifs should start on a new line.

File: Tiles/TransferAssemblerTile.cs [hunk 5]
```diff
@@ -175,6 +178,63 @@ private bool SearchStation(Recipe recipe, int x, int y, out bool alchemy)
...
+                        if (hashes[c] == entity.selectedRecipeHash)
+                        {
+                            index = c;
+                            break;
+                        }
+
+                    if (index == -1)
+                    {
+                        entity.selectedRecipeHash = 0;
+                        goto StartScrollHandling;
```
Comment (DRKV333) [inline Tiles/TransferAssemblerTile.cs]: Please don't use goto.

**Review by DRKV333 — CHANGES_REQUESTED** [review]
Overall, I like the idea, but the implementation could be improved a bit.

There are a few bugs. Scrolling on an assembler with no item set causes an exception. Sometimes the selection seems to scroll backwards (so scrolling up moves the selection down).

The recipe selection should not be shown every time an assembler is hovered, it takes up too much space. Also, I'm not sure if having it always active could cause someone to accidentally change recipes, while trying to scroll their hotbar.
So, the recipe selection should probably only be shown, once someone starts scrolling. Or, only show and activate the whole scrolling feature, once an item is right-clicked into the assembler, to make sure they meant it.
Also, either way, scrolling should be suppressed for other parts of the game UI, while the recipe selection is active.

I'm not entirely sold on the recipe hashing. Wouldn't it be easier to just store the actual recipe, serialize the result and ingredients on save, and find the corresponding real recipe from the loaded ones on load?

The whole architecture could be better, the `SyncSpecificData` part and the way this new `AssemblesHoverUI` copies a lot of things from the original `FilterHoverUI`, is a bit messy. That is partially my fault for not making this system generic enough. Some things could be factored out into base classes, but if you don't want to mess with that, this is fine by me too.

Some formatting issues should be fixed, see comments for examples.

**Comment by Ved-s** [comment]
I messed up branches at my repo a bit, but I think, I resolved it

**Comment by Ved-s** [comment]
Packet routing in the mod needs to be completely reworked, I can't send any additional data from AssemblerTE without fully overriding FilterTE's SyncData code.
I suggest registering Net Handlers with string IDs, so SyncableTE<T> can be made (If i make it now, AddHandler requires TE type, in some cases it's impossible to add handler to AssemblerTE which inherits from FilterTE and SyncableTE)

**Comment by Ved-s** [comment]
Frick it, i hate visual studio

**Comment by DRKV333** [comment]
I can also recommend VSCode, if you prefer a lighter editor.
I don't have the time right now to work on this mod, and fix things in your PR, but I hope we can pick this up later.

**Pull request closed without merging**